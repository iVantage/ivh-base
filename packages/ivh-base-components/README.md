
# ivh-base-components

> Dead simple React components for iVantage.

Check [this pen](https://codepen.io/jtrussell/pen/VWyGzL) out for a simple demo
and playground.


## Usage

Install the library:

```
yarn add @ivantagehealth/ivh-base-components
```

Import the modules you want to use and go to town:

```
import { AuAlert } from '@ivantagehealth/ivh-base-components'

// ...

<AuAlert primary>
  An alert!
</AuAlert>
```

Note - you may also include the iife build directly on our page from the
iVantage CDN.


### Accordions

We provide four atomic components for building accordions and expand/collapse
shelf experiences.

```
import {
  AuAccordion,
  AuAccordionPanel,
  AuAccordionHeader,
  AuAccordionBody
} from '@ivantage/ivh-base-components'

const Cmp = () => (
  <AuAccordion>
    <AuAccordionPanel active={true}>
      <AuAccordionHeader>
        Click to Open First Drawer!
      </AuAccordionHeader>
      <AuAccordionBody>
        Now you see me!
      </AuAccordionBody>
    </AuAccordionPanel>
    <AuAccordionPanel>
      <AuAccordionHeader>
        Click to Open Second Drawer!
      </AuAccordionHeader>
      <AuAccordionBody>
        Now you see me!
      </AuAccordionBody>
    </AuAccordionPanel>
  </AuAccordion>
)
```

Accordions are made up of panels which each have a header and body. Panels
accept and `active` prop to indicate whether they should be expanded (`active
=== true`) or collapsed. Note that it is up to you to enforce an "only one open
at a time" guarantee if that's what you want.


### Badges

Badges allow you to display "notification counts" in the upper right corner of
an element. Note that the containing element must have non-static positioning.
Some components provide a helper props ("bader") to support this, in other cases
you may use the `AuWithBadge` helper component to provide this positioning.

```
import {
  AuButton,
  AuBadge,
  AuWithBadge
} from '@ivantage/ivh-base-components'

// AuButton has the "badge" helper prop
const Cmp1 = () => (
  <AuButton badge>
    Clicky
    <AuBadge primary>5</AuBadge>
  </AuButton>
)

// The span can now have a badge
const Cmp2 = () => (
  <AuWithBadge>
    <span>
      Everything is crashing!
      <AuBadge danger>!!!</AuBadge>
    </span>
  </AuWithBadge>
)
```

Badges support the usual semantic props.

### Loading Masks

Loading masks provide a transparent, grayed out overlay for content that should
not be interacted with. This package includes a molecule which utilizes our
standard "spinner" dots with optional text:

```
import { LoadingMask } from '@ivantage/ivh-base-components'

// A simple loading mask with just dots and no text
const loadingThing = () => (
  <div style={{position: 'relative'}}>
    <LoadingMask />
  </div>
)

// A loading mask with awesome text
const otherLoadingThings = () => (
  <div style={{position: 'relative'}}>
    <LoadingMask>
      Please wait! Awesome things are happening :).
    </LoadingMask>
  </div>
)
```

Note that the mask will cover the first ancestor element with non-static
position.


### More

For more details and a complete list of availabe components, see:

- [Atoms in docs/atoms.md](docs/atoms.md)
- [Molecules in docs/molecules.md](docs/molecules.md)


## Conventions

### Atoms

- Atoms live in the `src/atoms` directory.
- Atoms should be automatically generated from the `atoms.json` file in the root
  of this repo. Do not create atoms by hand, they will be overwritten.
- Do not add tests for atoms. Atoms are essentially styled base elements (i.e.
  `div`, `button`, `input`, etc.)
- Generally speaking atom component names are the pascal case (PascalCase)
  version of their corresponding `ivh-base-style` classname roots. For example
  the `AuAlert` atom corresponds to elements with the css classnames
  `au-alert-*`. Secondary classnames are derived by atom props, e.g.:
  ```
  <AuAlert primary>Foo!</AuAlert>

  // translates to

  <p class="au-alert au-alert-primary">Foo!</p>
  ```
- You can change the base element using the `cmp` prop. All atoms support this:
  ```
  <AuAlert cmp="div" primary>Foo!</AuAlert>

  // translates to

  <div class="au-alert au-alert-primary">Foo!</div>
  ```
  Note that you *can* provide complex components here... it just may not be the
  best idea ;).
- Other props are forwarded along to the underlying DOM node (event handlers,
  input types, everything).


### Molecules

- Molecules live in the `src/molecules` directory.
- Molecules should not require custom CSS above what atoms provide.
  - When they do, be sure to use the variables provided by `ivh-base-style`.
- Molecules should automatically forward props such as `primary`, `success`,
  etc. to their underlying atoms where appropriate.
- Wherever possible, molecules should provide namespaces for prop forwarding:
  ```
  // Molcule has an inner `AuAtom` and will forward `foo={thingy}` to it. This
  // is meant as an emergency escape hatch and should not be abused (an
  // intentionaly leaky abstraction.
  <Molecule auAtomFoo={thingy} />
  ```

### Organisms

- Organisms live in the `src/organisms` directory.
- Organisms may require custom CSS and structure above what is provided by atoms
  and modules.
- Organisms are largely self-contained and often require little customization.


## Scripts

- `npm run atoms` - Build atomic components from `atoms.json`
- `npm run bump` Increment the package version and publish (git tag). I.e. `npm
  run bump minor`.
- `npm run clean` Kill old build artifacts.
- `npm test` Run all unit tests once. Test coverage results will be placed in
  `./coverage/<platform>`.
- `npm run test:watch` Execute tests continuously as files are updated.
- `npm run build` Clean old build artifacts and run webpack for production.
