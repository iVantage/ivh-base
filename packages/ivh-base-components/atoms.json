{
  "atoms": [
    [
      "AuBadge",
      "span",
      ["au-badge"],
      [],
      true,
      [],
      "An indicator for notification counts, etc."
    ],

    [
      "AuWithBadge",
      "span",
      ["au-with-badge"],
      [],
      false,
      [],
      "A helper for `AuBadge`, marks an element as having a badge."
    ],

    [
      "AuAccordion",
      "div",
      ["au-accordion"],
      [],
      false,
      [],
      "A top level accordion wrapper. Pair with panels, header, and body."
    ],

    [
      "AuAccordionPanel",
      "div",
      ["au-accordion-panel"],
      ["active"],
      false,
      [],
      "An accordion panel. Panels have a header and a body."
    ],

    [
      "AuAccordionHeader",
      "div",
      ["au-accordion-header"],
      [],
      false,
      [],
      "An accordion panel header."
    ],

    [
      "AuAccordionBody",
      "div",
      ["au-accordion-body"],
      [],
      false,
      [],
      "An accordion panel body."
    ],

    [
      "AuAlert",
      "p",
      ["au-alert"],
      ["flat"],
      true,
      [],
      "Fancy alerts for fancy people."
    ],

    [
      "AuButton",
      "button",
      ["au-btn"],
      ["secondary", "tertiary", "cta", "actions", "sm", "area", "link", "text", "badge"],
      true,
      [],
      "Fancy buttons for fancy people."
    ],

    [
      "AuCard",
      "div",
      ["au-card"],
      ["muted"],
      true,
      [],
      "A generic container."
    ],

    [
      "AuCardHeader",
      "div",
      ["au-card-header"],
      [],
      false,
      [],
      "A header for cards, not to be confused with `h` elements."
    ],

    [
      "AuClose",
      "span",
      ["au-close"],
      [],
      false,
      [],
      "A close button for cards, alerts, etc."
    ],

    [
      "AuDropdown",
      "div",
      ["au-dropdown"],
      ["open"],
      false,
      [],
      "An outer wrapper for dropdown components."
    ],

    [
      "AuDropdownButton",
      "div",
      ["au-dropdown-button"],
      [],
      false,
      [],
      "A button for dropdowns, this is what you would click to open a dropdown."
    ],

    [
      "AuDropdownBackdrop",
      "div",
      ["au-dropdown-backdrop"],
      [],
      false,
      [],
      "A transparent shim to listen for clicks to close a dropdown."
    ],

    [
      "AuDropdownMenu",
      "div",
      ["au-dropdown-menu"],
      [],
      false,
      [],
      "A menu for dropdowns."
    ],

    [
      "AuInput",
      "input",
      ["au-input"],
      [],
      false,
      [],
      "Fancy inputs for fancy people. In most cases may be used as a drop-in for `input`."
    ],

    [
      "AuCheckbox",
      "input",
      ["au-checkbox"],
      [],
      false,
      ["type='checkbox'"],
      "A special case for checkbox inputs."
    ],

    [
      "AuRadio",
      "input",
      ["au-radio"],
      [],
      false,
      ["type='radio'"],
      "A special case for radio button inputs."
    ],

    [
      "AuList",
      "ul",
      ["au-list"],
      ["slim", "relaxed", "selectable"],
      false,
      [],
      "A list wrapper."
    ],

    [
      "AuListItem",
      "li",
      ["au-list-item"],
      ["active"],
      false,
      [],
      "A list item."
    ],

    [
      "AuMarker",
      "span",
      ["au-marker"],
      ["placeholder", "muted"],
      true,
      [],
      "A color marker, often used to connect legends with data visualizations."
    ],

    [
      "AuPill",
      "span",
      ["au-pill"],
      ["active"],
      true,
      [],
      "Fancy pills for fancy people."
    ],

    [
      "AuPillNav",
      "a",
      ["au-pill-nav", "au-pill"],
      ["active"],
      false,
      [],
      "A pill styled for navigation."
    ],

    [
      "AuLabel",
      "span",
      ["au-label"],
      ["active"],
      true,
      [],
      "A label for form inputs."
    ],

    [
      "AuShadow",
      "div",
      ["au-shadow"],
      ["short", "medium", "tall"],
      false,
      [],
      "Apply a shadown to give the wrapped element some depth."
    ],

    [
      "AuTable",
      "table",
      ["au-table"],
      ["slim", "relaxed", "striped", "borderless"],
      true,
      [],
      "Fancy tables for fancy people. Pair with table rows, columns, and/or cells."
    ],

    [
      "AuTableRow",
      "tr",
      ["au-table-row"],
      [],
      false,
      [],
      "Fancy table rows for fancy people."
    ],

    [
      "AuTableCol",
      "td",
      ["au-col"],
      ["checkbox", "text", "data"],
      false,
      [],
      "Fancy table columns for fancy people."
    ],

    [
      "AuGrid",
      "div",
      ["au-grid"],
      [],
      true,
      [],
      "A grid, similar to tables but with divs and such instead of an actual `table` element."
    ],

    [
      "AuCell",
      "div",
      ["au-cell"],
      ["divider", "row", "checkbox", "text", "data"],
      true,
      [],
      "A cell for grids."
    ],

    [
      "AuTextarea",
      "textarea",
      ["au-input"],
      [],
      false,
      [],
      "A special case for textarea inputs."
    ],

    [
      "AuTabContainer",
      "ul",
      ["au-tab-container"],
      [],
      false,
      [],
      "A simple tab container."
    ],

    [
      "AuTab",
      "li",
      ["au-tab"],
      [],
      false,
      [],
      "A simple tab."
    ],

    [
      "AuTabContent",
      "div",
      ["au-tab-content"],
      [],
      false,
      [],
      "A panel for contents of a tab."
    ],

    [
      "AuModal",
      "div",
      ["au-modal"],
      [],
      true,
      [],
      "A dialog for notifications."
    ],

    [
      "AuModalHeader",
      "div",
      ["au-modal-header"],
      [],
      false,
      [],
      "The header of a modal."
    ],

    [
      "AuModalBody",
      "div",
      ["au-modal-body"],
      [],
      false,
      [],
      "The body of a modal."
    ],

    [
      "AuModalFooter",
      "div",
      ["au-modal-footer"],
      [],
      false,
      [],
      "The footer of a modal."
    ],

    [
      "AuModalBackdrop",
      "div",
      ["au-modal-backdrop"],
      [],
      false,
      [],
      "Backdrop of a modal. `AuModal` components should generally be nested in a backdrop."
    ]
  ]
}
