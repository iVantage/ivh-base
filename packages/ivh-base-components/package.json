{
  "name": "@ivantagehealth/ivh-base-components",
  "version": "5.7.0",
  "description": "Dead simple React components for iVantage.",
  "main": "dist/ivh-base-components.js",
  "files": [
    "dist"
  ],
  "scripts": {
    "atoms": "node scripts/atoms",
    "build": "yarn run clean && yarn run atoms && yarn run docs && yarn run build:molecules",
    "build:molecules": "cross-env NODE_ENV=production webpack",
    "build:watch": "yarn run clean && cross-env NODE_ENV=development webpack --watch",
    "clean": "rimraf dist",
    "docs": "node scripts/docs",
    "lint": "standard --fix && yarn run lint:free",
    "lint:free": "echo Such style. Much class. Wow.",
    "test": "jest",
    "test:watch": "jest --watchAll"
  },
  "author": "iVantage",
  "license": "UNLICENSED",
  "jest": {
    "moduleNameMapper": {
      "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/package.json",
      "\\.css$": "<rootDir>/package.json"
    }
  },
  "devDependencies": {
    "@babel/core": "7.6.0",
    "autoprefixer": "^8.1.0",
    "babel-eslint": "^10.0.3",
    "babel-jest": "^24.9.0",
    "babel-loader": "^8.0.6",
    "babel-preset-react-app": "^9.0.2",
    "babel-runtime": "^6.26.0",
    "cross-env": "^5.0.0",
    "css-loader": "^2.1.1",
    "file-loader": "^3.0.1",
    "jest": "^24.9.0",
    "postcss-loader": "^3.0.0",
    "react": "^16.9.0",
    "react-dom": "^16.9.0",
    "react-test-renderer": "^16.9.0",
    "rimraf": "^3.0.0",
    "standard": "11.0.0",
    "style-loader": "^1.0.0",
    "webpack": "^4.40.2",
    "webpack-cli": "^3.3.9"
  },
  "standard": {
    "ignore": [
      "dist"
    ],
    "parser": "babel-eslint"
  }
}
