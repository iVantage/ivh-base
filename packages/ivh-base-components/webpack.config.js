/* eslint-env node */

const path = require('path')
const srcPath = path.join(__dirname, 'src')
const distPath = path.join(__dirname, 'dist')

module.exports = {
  mode: 'development', // process.env.NODE_ENV,
  bail: true,
  entry: ['./src/index.js'],
  resolve: {
    extensions: ['.js']
  },
  output: {
    library: 'IvhBaseComponents',
    libraryTarget: 'umd',
    path: distPath,
    filename: 'ivh-base-components.js'
  },
  externals: {
    react: {
      root: 'React',
      commonjs: 'react',
      commonjs2: 'react',
      amd: 'react'
    },
    'react-dom': {
      root: 'ReactDOM',
      commonjs: 'react-dom',
      commonjs2: 'react-dom',
      amd: 'react-dom'
    },
    '@ivantagehealth/ivh-base-components': '@ivantagehealth/ivh-base-components'
  },
  module: {
    rules: [
      {
        oneOf: [
          {
            test: /\.js$/,
            loader: 'babel-loader'
          },
          {
            test: /\.css$/,
            use: [
              'style-loader',
              {
                loader: 'css-loader',
                options: {
                  importLoaders: 1
                }
              },
              'postcss-loader'
            ]
          }
        ]
      }
    ]
  }
}
