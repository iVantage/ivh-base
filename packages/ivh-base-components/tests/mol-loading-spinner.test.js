import React from 'react'
import renderer from 'react-test-renderer'
import { AuLoadingMask } from '../dist/ivh-base-components'

test('AuLoadingMask should render consistently', () => {
  const tree = renderer
    .create(
      <AuLoadingMask>
        Please wait, the bears are waking from hibernation...
      </AuLoadingMask>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
