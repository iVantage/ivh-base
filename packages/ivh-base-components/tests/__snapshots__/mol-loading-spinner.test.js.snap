// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuLoadingMask should render consistently 1`] = `
<div
  className="au-mask"
>
  <div
    className="au-mask-spinner"
  >
    <span
      className="au-spinner-dot"
    />
    <span
      className="au-spinner-dot"
    />
    <span
      className="au-spinner-dot"
    />
  </div>
  Please wait, the bears are waking from hibernation...
</div>
`;
